### PR TITLE
Multi select quick match

### PR DESCRIPTION
**Pull request by Undergrid**
Adds a feature to allow the user to batch Quick Match the selected audiobooks.

A button is added to the Appbar when one or more audiobooks are selected using the multi-select option (left most button on the below image).

button

Once clicked, a dialog is shown allowing the user to select the metadata provider and to select whether exiting covers or metadata should be overwritten.

dialog

When "Continue" is selected, a quick match is run against each selected audiobook, updating metadata and covers as required.

### CODE REVIEW

**Comment by advplyr** [comment]
Awesome. I think we should initialize the provider to whatever the default provider is for the library.

There are some getters in `libraries.js` to get a library provider or get the current library.

Also, we can show the user how many items were updated and how many items were unable to find a match.
Currently the `quickMatchLibraryItem` function is returning a warning if it doesn't find a match. https://github.com/advplyr/audiobookshelf/blob/master/server/scanner/Scanner.js#L695

**Comment by Undergrid** [comment]
How would you like to show the updated and unmatched count?  Would in the toast be okay?

**Comment by advplyr** [comment]
Yeah that works, the toast message can have line breaks `\n`.

Eventually we will probably have to change this so that the API request doesn't need to wait for the matches to finish and instead gets the result from a socket event. The reason for that is because some people have very slow systems and the API request will timeout before it finishes.

**Comment by Undergrid** [comment]
If you can point me at an example of doing that, I can probably get it working that way.  Might as well do it properly from the start.

**Comment by skyzuma** [comment]
is possible to select details if the switch goes to on, like (update only) language, genre, ... (for selected) ?

**Comment by advplyr** [comment]
> If you can point me at an example of doing that, I can probably get it working that way. Might as well do it properly from the start.

I don't have a good example to point to yet for this. We would need to set some temporary id locally for the quick match request and setup a socket event listener (something like `batch_quickmatch_complete`). Once finished the backend emits that event including the temporary id so know it is this clients request that was completed.
Most of the socket event listeners are set in the `default.vue` layout.
If you want to go for this that would be great but up to you.

**Comment by advplyr** [comment]
> is possible to select details if the switch goes to on, like (update only) language, genre, ... (for selected) ?
> 
> li > switch > details ul > swicth > language ul > switch > genre

We don't have support for this in the backend and would be a lot of re-working that is out of the scope of this PR

**Comment by Undergrid** [comment]
So I'm going to look at the websocket part, but in the mean time I've set the provider and added the counts to the toast.  The toast looks a little different depending on the counts:

match
nomatch
matchunmatch
nomatchunmatch

**Comment by Undergrid** [comment]
@advplyr PR has now been updated as request, including with the websocket feedback etc.

File: client/components/modals/BatchQuickMatchModel.vue [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+          this.$store.commit('setProcessingBatch', false)
+          this.show = false
+        })
+        .catch((error) => {
+          this.$toast.error('Batch quick match failed')
+          console.error('Failed to batch quick match', error)
+          this.processing = false
+          this.$store.commit('setProcessingBatch', false)
+          this.show = false
+        })
```
Comment (advplyr) [inline client/components/modals/BatchQuickMatchModel.vue:139]:
```suggestion
        .then(() => {
          this.$toast.info('Batch quick match of ' + this.selectedBookIds.length + ' books started!')
        }).catch((error) => {
          this.$toast.error('Batch quick match failed')
          console.error('Failed to batch quick match', error)
        }).finally(() => {
          this.processing = false
          this.$store.commit('setProcessingBatch', false)
          this.show = false
        })
```

File: server/controllers/LibraryItemController.js [hunk 2]
```diff
@@ -305,6 +305,42 @@ class LibraryItemController {
...
+    var itemsUpdated = 0
+    var itemsUnmatched = 0
+
+    var matchData = req.body
+    var options = matchData.options || {}
+    var items = matchData.libraryItemIds
+    if (!items || !items.length) {
+      return res.sendStatus(500)
+    }
+    res.sendStatus(200);
```
Comment (advplyr) [inline server/controllers/LibraryItemController.js:324]:
```suggestion
    res.sendStatus(200)
```

File: server/controllers/LibraryItemController.js [hunk 3]
```diff
@@ -305,6 +305,42 @@ class LibraryItemController {
...
+        } else if (matchResult.warning) {
+            itemsUnmatched++
+        }
+    }
+    
+    var result = {
+      success: itemsUpdated > 0,
+      updates: itemsUpdated,
+      unmatched: itemsUnmatched
+    };
```
Comment (advplyr) [inline server/controllers/LibraryItemController.js:340]:
```suggestion
    }
```

File: client/components/modals/BatchQuickMatchModel.vue [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      lastUsedLibrary: undefined,
+      options: {
+        provider: undefined,
+        overrideDetails: true,
+        overrideCover: true,
+        overrideDefaults: true
+      },
+      tooltips: {
+          updateCovers: 'Allow overwriting of existing covers for the selected books when a match is located.',
+          updateDetails: 'Allow overwriting of existing details for the selected books when a match is located.'
```
Comment (advplyr) [inline client/components/modals/BatchQuickMatchModel.vue:67]:
```suggestion
        updateCovers: 'Allow overwriting of existing covers for the selected books when a match is located.',
        updateDetails: 'Allow overwriting of existing details for the selected books when a match is located.'
```

File: client/components/modals/BatchQuickMatchModel.vue [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        <div class="py-4 px-4">
+          <h1 class="text-2xl">Quick Match {{ selectedBookIds.length }} Books</h1>
+        </div>
+        
+        <div class="w-full overflow-y-auto overflow-x-hidden max-h-96">
+          <div class="flex px-8 items-center py-2">
+            <p class="pr-4">Provider</p>
+            <ui-dropdown v-model="options.provider" :items="providers" small />
+          </div>
+          <p class="text-base px-8 py-2">Quick Match will attempt to add missing covers and metadata for the selected books.  Enable the options below to allow Quick Match to overwrite existing covers and/or metadata.</p>
```
Comment (advplyr) [inline client/components/modals/BatchQuickMatchModel.vue:20]:
```suggestion
          <p class="text-base px-8 py-2">Quick Match will attempt to add missing covers and metadata for the selected books. Enable the options below to allow Quick Match to overwrite existing covers and/or metadata.</p>
```

File: client/components/modals/BatchQuickMatchModel.vue [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    currentLibraryId() {
+      return this.$store.state.libraries.currentLibraryId
+    },
+    providers() {
+      if (this.isPodcast) return this.$store.state.scanners.podcastProviders
+      return this.$store.state.scanners.providers
+    },
+    libraryProvider() {
+      return this.$store.getters['libraries/getLibraryProvider'](this.currentLibraryId) || 'google'
+    },
```
Comment (advplyr) [inline client/components/modals/BatchQuickMatchModel.vue:105]:
```suggestion
    }
```

**Review by advplyr — CHANGES_REQUESTED** [review]

**Comment by advplyr** [comment]
Works beautifully, thanks!

**Pull request merged**